File: Diagrammatize.Design.Pattern/code/cxx/State/A3/DayState.cpp
#include "DayState.hpp"
#include "NightState.hpp"
#include "NoonState.hpp"
#include "Context.hpp"

DayState::DayState() {
}

State* DayState::getInstance() {
    static DayState singleton;
    return &singleton;
}

void DayState::doClock(Context* context, int hour) {
    if (hour < 9 || 17 <= hour) {
        context->changeState(NightState::getInstance());
    } else if (12 <= hour && hour < 13) {                   
        context->changeState(NoonState::getInstance());       
    }                                                       
}

void DayState::doUse(Context* context) {
    context->recordLog("使用金库(白天)");
}

void DayState::doAlarm(Context* context) {
    context->callSecurityCenter("按下警铃(白天)");
}

void DayState::doPhone(Context* context) {
    context->callSecurityCenter("正常通话(白天)");
}

std::string DayState::toString() const {
    return "[白天]";
}

File: Diagrammatize.Design.Pattern/code/python/State/A3/DayState.py
from State import State

class DayState(State):
    _singleton = None

    def __init__(self):
        print("create DayState")
        pass

    @classmethod
    def getInstance(cls):
        if not cls._singleton:
            cls._singleton = cls()
        return cls._singleton

    def doClock(self, context, hour):
        from NightState import NightState
        from NoonState import NoonState
        if hour < 9 or 17 <= hour:
            context.changeState(NightState.getInstance())
        elif 12 <= hour < 13:
            context.changeState(NoonState.getInstance())

    def doUse(self, context):
        context.recordLog("使用金库(白天)")

    def doAlarm(self, context):
        context.callSecurityCenter("按下警铃(白天)")

    def doPhone(self, context):
        context.callSecurityCenter("正常通话(白天)")

    def __str__(self):
        return "[白天]"

